### PR TITLE
KAFKA-7146: Grouping consumer requests per consumer coordinator in admin client in describeConsumerGroups

### PR DESCRIPTION
**Pull request by shunge**
This is a subtask from KAFKA-6788 (https://issues.apache.org/jira/browse/KAFKA-6788)

Basically the idea is to use thread-safe collections like ConcurrentLinkedQueue and ConcurrentHashMap to save to result of each "findCoordinator" response. In the map, we will have the coordinator node Id as the key, and the value will be the collection of all the belonging groupIds. 

When all the groupId processed (either successful or completed exceptionally), then we will iterate through the map and group the requests by coordinators.

This PR is mainly for discussion, there are a couple concerns:

1. Code logic: Is there any thing wrong with the logic of this solution?
2. Concurrency issues: can these collections guarantee the thread-safety?
3. Memory issue: using too many collections to store information, should cut down the memory usage.
4. Efficiency: is this more efficient than the implementation before (i.e. latency)?
5. Naming issues.
6. More unit tests.

I am open to all kind of suggestions and feedbacks, and will make several other commits after I receive opinion. 

P.S. Is a KIP needed?

@guozhangwang @cmccabe Thank you!

### CODE REVIEW

**Comment by shunge** [comment]
Hi, when you guys have the time, would you mind take a look at this? If you guys think I should create an KIP to explain, I will then do that. Thanks! @guozhangwang @cmccabe @mjsax

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (lambdaliu) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I suggest to use HashMap and LinkedList here since only one thread will process all responses.

Comment (shunge) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Hi, @lambdaliu , I was assuming the process of KafkaFuture will be asynchronous and multi-thread, what do you think?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (andrasbeni) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: nit: unnecessary whitespaces

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (andrasbeni) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: This comment should now be removed

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (andrasbeni) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: As I understand it, handleResponse will always be called by AdminClientRunnable from the single 'network thread' (KafkaAdminClient.thread).

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (andrasbeni) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: This method sends multiple requests now. It would be useful to rename it to `sendGroupRequests()`

**Review by andrasbeni — CHANGES_REQUESTED** [review]
Thanks for the patch.
I left a few comments.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (shunge) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I see, I will modify it then. Thank you @andrasbeni @lambdaliu

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (shunge) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Fixed.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (shunge) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Fixed.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (shunge) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Fixed.

**Comment by guozhangwang** [comment]
ping @cmccabe for reviews.

**Comment by shunge** [comment]
@cmccabe @guozhangwang when you guys got the chance, could you guys review this? Thanks!

**Comment by guozhangwang** [comment]
@shunge Here are my general thoughts:

a). I think the general idea is on the right track: basically for the group requests, we have two phases: find the coordinator and then send the actual request to the coordinator. So in order to group the requests sending to the same coordinator for the second phase we should keep track of which groups are belonging to the same coordinator, in a data structure.

BTW I think we do not need a thread-safe structure, since the admin client only has a single network thread.

b) Memory should not be a big issue, but if you do want to optimize a bit, I think we can also merge some of your added maps here. For example `fcResponses` is used for returning the group coordinator in the final `ConsumerGroupDescription` constructor, but you can get the same information from `coordinators`: just loop through the contained list of group Ids until you found the id. Similarly, `returnedGroupId` is only used to check if all group's coordinators have been found, this can also be done via `coordinators` by summing up the sizes of each entry's value.

c) Note that in admin client the network thread is simply working in a loop of: reading for responses from sockets, checking if new requests are pushed from caller thread and send them. Therefore, one thing you'd need to be careful is that, now since we've separated these two phases, we are actually introducing a implicit synchronization barrier (the `returnedGroupId` is used for this purpose), that we would only start the second phase until all group's coordinator have found or have failed the findCoordinator request. Before that, for an individual group, we would send out the request immediately when we know it's coordinator. Now we will be blocked until the slowest findCoordinator returns. So we may need to consider doing some "partial grouping", that is, once we have received some findCoordinator responses, and updated the `coordinators` map, we check how much time has passed and `if it is already late` (some ad hoc heuristics, maybe?), we just send whatever is grouped so far and do not wait for the remaining, which would then have to been sent via another request round-trip later.

d) A final meta comment, as has been mentioned in the JIRA: `I think this would be a hard optimization to do correctly because of the error handling issues. If you get an error for some, but not all, elements of the batch, you want to retry just those elements.`. So just make sure that when some error was received from the grouped response, we need to separate which groups has error from others which are fine (e.g. in `DescribeGroupsResponse` our schema contains a per-group error code), and then handling the error. Note that we do not need to consider the retry logic in your PR as it is for a different JIRA: https://issues.apache.org/jira/browse/KAFKA-6789 so just follow the current logic and not worry about retrying is okay.

Finally, really sorry for the late replies... we are in deep lack of active committers to review PRs compares to our num. active contributors and num. open PRs filed every day. 

I'd leave to @cmccabe to give some detailed comments.

**Comment by shunge** [comment]
@guozhangwang Please do not worry about the the timing of your replies! I am the one who should be saying thanks, it is more like an opportunity for me to learn than anything else. 
I totally understand you guys have a tremendous amount of works that need to be done! I will review your comment closely and think about it. Thank you~

**Comment by cmccabe** [comment]
Sorry that I didn't comment earlier.  I meant to look at this but I kept getting distracted by something else.

One issue that I see here (and I think @guozhangwang also pointed it out) is that this may actually reduce performance under some situations... at least in terms of latency.  Maybe we should have a KIP to enable batching in FindCoordinator before trying to optimize this.

**Comment by shunge** [comment]
Hi @cmccabe , I totally agree with what you and @guozhangwang said. I think I will start a KIP on this one. Thanks!

**Comment by shunge** [comment]
Hi @cmccabe , when you said enable batching in FindCoordinator, do you mean enable batching in FindCoordinatorRequest class? Or you still mean to modify and improve on what we currently have? Thanks!

**Comment by guozhangwang** [comment]
@shunge The proposal is that, we extend FindCoordinatorRequest to have multiple consumer ids: today each FindCoordinatorRequest only contains a single consumer id, so in our scenario we need to send N request for N consumer groups still. If we can request for coordinators in a single request, then the workflow could be simplified to:

1) send a single FindCoordinatorRequest to a broker asking for coordinators of all consumer groups.
1.a) note that the response may still succeed in finding some coordinators while error on others, and we need to handle them on that granularity (see below).
2) and then for the collected coordinator, group them by coordinator id and send one request per coordinator destination.

Note that this change would require the version to be bumped up, to `FIND_COORDINATOR_REQUEST_V3` for such protocol changes, also the RESPONSE version should be bumped up in order to include multiple coordinators.

**Comment by shunge** [comment]
Hi @guozhangwang , I have been staring at the FindCoordinatorRequest.java and FindCoordinatorResponse for a while. But I don't think I understand Kafka's architect enough to make this KIP. Where do you suggests me to look in order for me to start? Any good documents? Thank!

**Comment by guozhangwang** [comment]
Besides these two classes, I'd suggest you taking a look at `KafkaApis.scala` file, which contains a bunch of `handleXXX` functions. The one you should be interested in is the `handleFindCoordinatorRequest` function, which parses a request into the `FindCoordinatorRequest` and at the end of the handling logic it will call `createResponse` to generate the `FindCoordinatorResponse` object and pass to `sendResponseMaybeThrottle` (you can ignore the `sendResponseMaybeThrottle` impl as it does not need to be modified, just assuming it does the sending of the response).

That logic handles one group id at a time, and the suggestion is to extend it to handle multiple group ids from the newer version of `handleFindCoordinatorRequest`.

**Comment by shunge** [comment]
@guozhangwang @cmccabe Thank you so much for your help!

So from what I gather so far, these are some changes that need to be make:

1. Update the `Schema` in `FindCoordinatorRequest` to store an array of groupID (with a new `FIND_COORDINATOR_REQUEST_V3`); modify the class structure to take more than one coordinatorKey (`List<String>` groupIds). Update the setters and getters of course.
2. Update `FindCoordinatorResponse` 's class to have a mapping structure (now it only has Node to support one result, we need to support multiple results so something like `<GroupId, CoordinatorNodeMetadata>` or `<CoordinatorNodeID, List<GroupId>>`)
3. Update `handleFindCoordinatorRequest` to correctly parse the new `FindCoordinatorRequest` and map the result to the new `FindCoordinatorResponse`.

What do you think? Thank you!

If this idea is close enough, I will then go ahead and create a KIP based on what we just discussed so other people can chip in as well.

**Comment by guozhangwang** [comment]
Yes that sounds good to me, a good study of the internals!

In 2) I think`<GroupId, CoordinatorNodeMetadata>` is better as it matches the request format.

**Pull request closed without merging**
